reframe/core/pipeline.py

```diff
@@ -1666,63 +1691,52 @@ def check_performance(self):
                         f"(class {self.__class__.__qualname__})"
                     )
 
-                # Build the performance dict
-                if not hasattr(self, '_rfm_perf_vars'):
-                    self.perf_patterns = {}
-                    for fn in self._rfm_perf_fns:
-                        self.perf_patterns[fn._rfm_perf_name] = fn(self)
-
-                else:
-                    self.perf_patterns = self._rfm_perf_vars()
-
                 # Log the performance variables
                 self._setup_perf_logging()
-                for tag, expr in self.perf_patterns.items():
+                for tag, expr in self.perf_variables.items():
                     try:
-                        value, unit = sn.evaluate(expr)
-                        unit = getattr(unit, '_rfm_perf_units', None)
-                        if not unit:
-                            raise TypeError('Trigger the TypeError below')
-
-                    except TypeError:
-                        raise ReframeSyntaxError(
-                            f'function assigned for performance variable '
-                            f'{tag!r} is not a performance function'
+                        value = expr.evaluate()
+                        unit = expr.unit
+                    except Exception as e:
+                        logging.getlogger().warning(
+                            f'skipping evaluation of performance variable '
+                            f'{tag!r} in test {self.name!r}: {e}'
                         )
+                        continue
 
-                    key = '%s:%s' % (self._current_partition.fullname, tag)
+                    key = f'{self._current_partition.fullname}:{tag}'
                     try:
                         ref = self.reference[key]
 
-                        # If units are also provided in the reference, make
-                        # sure they match with the units provided by the
-                        # performance function.
+                        # If units are also provided in the reference, raise
+                        # a warning if they match with the units provided by
+                        # the performance function.
                         if len(ref) == 4:
                             if ref[3] != unit:
-                                raise ReframeSyntaxError(
-                                    f'units for {tag!r} in the reference '
-                                    f'{key!r} do not match those specified in '
-                                    f'the performance function ({unit})'
+                                logging.getlogger().warning(
+                                    f'reference unit ({key!r}) for the '
+                                    f'performance variable {tag!r} '
+                                    f'does not match the unit specified '
+                                    f'in the performance function ({unit!r})'
```

Comment:
Perhaps we need to say which one of the two will be ignored. We can simply add that "{unit!r} will be used"